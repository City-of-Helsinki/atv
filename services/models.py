from django.db import models
from django.utils.translation import gettext_lazy as _


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated at"), auto_now=True)

    class Meta:
        abstract = True


class Service(TimestampedModel):
    name = models.CharField(verbose_name=_("name"), max_length=50)
    description = models.TextField(verbose_name=_("description"), blank=True)

    class Meta:
        verbose_name = _("service")
        verbose_name_plural = _("services")

    def __str__(self):
        return self.name
