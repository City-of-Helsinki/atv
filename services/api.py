from rest_framework import status, viewsets
from rest_framework.response import Response


class DocumentsViewSet(viewsets.ViewSet):
    def create(self, request):
        return Response({"status": "CREATED"}, status=status.HTTP_201_CREATED)
