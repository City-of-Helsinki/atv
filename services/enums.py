from django.db.models import TextChoices


class ServicePermissions(TextChoices):
    ADD_DOCUMENTS = "can_add_documents", "Can add documents"
    CHANGE_DOCUMENTS = "can_change_documents", "Can change documents"
    DELETE_DOCUMENTS = "can_delete_documents", "Can delete documents"
    VIEW_DOCUMENTS = "can_view_documents", "Can view documents"
    ADD_ATTACHMENTS = "can_add_attachments", "Can add attachments"
    CHANGE_ATTACHMENTS = "can_change_attachments", "Can change attachments"
    DELETE_ATTACHMENTS = "can_delete_attachments", "Can delete attachments"
    VIEW_ATTACHMENTS = "can_view_attachments", "Can view attachments"
