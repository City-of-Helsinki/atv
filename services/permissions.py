from rest_framework_api_key.permissions import BaseHasAPIKey

from .models import ServiceAPIKey


class HasServiceAPIKey(BaseHasAPIKey):
    model = ServiceAPIKey
