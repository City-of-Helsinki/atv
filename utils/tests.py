def assert_in_errors(message, response: dict):
    assert "errors" in response
    assert str(message) in response.get("errors")
