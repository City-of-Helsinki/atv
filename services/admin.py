from django.contrib import admin
from django.db.models import Count
from django.template.defaultfilters import truncatechars
from django.utils.translation import gettext_lazy as _
from guardian.admin import GuardedModelAdmin
from rest_framework_api_key.admin import APIKeyModelAdmin
from rest_framework_api_key.models import APIKey

from .models import Service, ServiceAPIKey


@admin.register(Service)
class ServiceAdmin(GuardedModelAdmin):
    list_display = ("name", "short_description", "api_key_count")
    search_fields = ("name",)

    def get_queryset(self, request):
        qs = super().get_queryset(request)
        return qs.annotate(api_keys_count=Count("api_keys"))

    @admin.display(description=_("description"))
    def short_description(self, obj: Service):
        return truncatechars(obj.description, 255)

    @admin.display(description=_("api keys"), ordering="api_keys_count")
    def api_key_count(self, obj):
        return obj.api_keys_count


@admin.register(ServiceAPIKey)
class ServiceAPIKeyModelAdmin(APIKeyModelAdmin):
    list_display = (*APIKeyModelAdmin.list_display, "get_service_name")
    search_fields = (*APIKeyModelAdmin.search_fields, "service__name")
    autocomplete_fields = ("service",)
    list_select_related = ("service",)

    @admin.display(description=_("service"), ordering="-service__name")
    def get_service_name(self, obj):
        return obj.service.name


admin.site.unregister(APIKey)
