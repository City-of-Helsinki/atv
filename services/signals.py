import logging

from django.contrib.auth import get_user_model
from django.contrib.auth.models import Group
from django.db.models.signals import post_delete, post_save
from django.dispatch import receiver
from django.utils.text import slugify
from guardian.shortcuts import assign_perm

from .models import Service, ServiceAPIKey

logger = logging.getLogger(__name__)


@receiver(post_save, sender=Service, dispatch_uid="create_service_group")
def create_service_group(sender, instance: Service, created: bool, **kwargs):
    if created:
        group, _created = Group.objects.get_or_create(name=instance.name)
        logger.debug(f"Created permission group for service: {instance.name}")
        for perm in Service._meta.permissions:
            assign_perm(f"{perm[0]}", group, instance)
            logger.debug(f"Added permission to group: {perm[0]}")


@receiver(post_delete, sender=Service, dispatch_uid="remove_service_group")
def remove_service_group(sender, instance: Service, **kwargs):
    Group.objects.filter(name=instance.name).delete()
    logger.debug(f"Removed permission group: {instance.name}")


@receiver(post_save, sender=ServiceAPIKey, dispatch_uid="create_service_api_key_user")
def create_service_api_key_user(
    sender, instance: ServiceAPIKey, created: bool, **kwargs
):
    if created and not instance.user:
        user_model = get_user_model()
        username = slugify(f"service-api-key-{instance.service.name}-{instance.name}")
        user = user_model.objects.create_user(username=username)
        instance.user = user
        instance.save()
        logger.debug(
            f"Created user {user} for ServiceAPIKey {instance} permission group for service: {instance.name}"
        )
        for perm in Service._meta.permissions:
            assign_perm(f"{perm[0]}", user, instance.service)
            logger.debug(f"Added permission to group: {perm[0]}")
