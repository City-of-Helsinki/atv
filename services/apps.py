from django.apps import AppConfig
from django.utils.translation import gettext_lazy as _


class ServicesConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "services"
    verbose_name = _("services")

    def ready(self):
        import services.signals  # noqa
