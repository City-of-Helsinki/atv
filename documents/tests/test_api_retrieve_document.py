import uuid

from freezegun import freeze_time
from guardian.shortcuts import assign_perm
from rest_framework import status
from rest_framework.reverse import reverse

from atv.tests.factories import GroupFactory
from documents.tests.factories import DocumentFactory
from services.enums import ServicePermissions


@freeze_time("2020-06-01T00:00:00Z")
def test_list_document_service_user(api_client, user, service, snapshot):
    expected_document_id = "485af718-d9d1-46b9-ad7b-33ea054126e3"
    expected_user_id = "66d0bfd0-308c-484d-aa22-301512899ae3"

    group = GroupFactory(name=service.name)
    user.groups.add(group)
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)

    document = DocumentFactory(
        id=expected_document_id,
        service=service,
        user__uuid=expected_user_id,
        tos_function_id="81eee139736f4e52b046a1b27211c202",
        tos_record_id="0f499febb6414e1dafc93febca5ef312",
        transaction_id="bd3fd958-cfeb-4ab1-bea4-5c058e8fee5c",
    )

    api_client.force_login(user=user)
    response = api_client.get(reverse("documents-detail", args=[document.id]))

    assert response.status_code == status.HTTP_200_OK

    body = response.json()
    assert body.pop("user_id") == str(expected_user_id)
    snapshot.assert_match(body)


@freeze_time("2020-06-01T00:00:00Z")
def test_retrieve_document_owner(api_client, user, service, snapshot):
    expected_document_id = "485af718-d9d1-46b9-ad7b-33ea054126e3"

    group = GroupFactory(name=service.name)
    user.groups.add(group)

    document = DocumentFactory(
        id=expected_document_id,
        service=service,
        user=user,
        tos_function_id="81eee139736f4e52b046a1b27211c202",
        tos_record_id="0f499febb6414e1dafc93febca5ef312",
        transaction_id="bd3fd958-cfeb-4ab1-bea4-5c058e8fee5c",
    )

    api_client.force_login(user=user)
    response = api_client.get(reverse("documents-detail", args=[document.id]))

    assert response.status_code == status.HTTP_200_OK

    body = response.json()
    assert body.pop("user_id") == str(user.uuid)
    snapshot.assert_match(body)


@freeze_time("2020-06-01T00:00:00Z")
def test_list_document_superuser(superuser_api_client, snapshot):
    expected_document_id = "485af718-d9d1-46b9-ad7b-33ea054126e3"
    expected_user_id = "66d0bfd0-308c-484d-aa22-301512899ae3"

    document = DocumentFactory(
        id=expected_document_id,
        user__uuid=expected_user_id,
        tos_function_id="81eee139736f4e52b046a1b27211c202",
        tos_record_id="0f499febb6414e1dafc93febca5ef312",
        transaction_id="bd3fd958-cfeb-4ab1-bea4-5c058e8fee5c",
    )

    response = superuser_api_client.get(reverse("documents-detail", args=[document.id]))

    assert response.status_code == status.HTTP_200_OK

    body = response.json()
    assert body.pop("user_id") == str(expected_user_id)
    snapshot.assert_match(body)


def test_list_document_no_permissions(api_client):
    response = api_client.get(reverse("documents-detail", args=[uuid.uuid4()]))

    body = response.json()
    assert response.status_code == status.HTTP_403_FORBIDDEN
    assert (
        body.get("detail", "") == "You do not have permission to perform this action."
    )
