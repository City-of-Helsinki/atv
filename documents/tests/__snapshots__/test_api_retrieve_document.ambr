# serializer version: 1
# name: test_list_document_service_user
  dict({
    'attachments': list([
    ]),
    'business_id': '',
    'content': dict({
    }),
    'content_schema_url': None,
    'created_at': '2020-06-01T03:00:00+03:00',
    'deletable': True,
    'delete_after': None,
    'document_language': None,
    'draft': False,
    'human_readable_type': dict({
    }),
    'id': '485af718-d9d1-46b9-ad7b-33ea054126e3',
    'locked_after': None,
    'metadata': dict({
    }),
    'service': 'service 215',
    'status': dict({
      'status_display_values': dict({
      }),
      'timestamp': '2020-06-01T03:00:00+03:00',
      'value': 'testing',
    }),
    'status_histories': list([
    ]),
    'tos_function_id': '81eee139736f4e52b046a1b27211c202',
    'tos_record_id': '0f499febb6414e1dafc93febca5ef312',
    'transaction_id': 'bd3fd958-cfeb-4ab1-bea4-5c058e8fee5c',
    'type': '',
    'updated_at': '2020-06-01T03:00:00+03:00',
  })
# ---
# name: test_list_document_superuser
  dict({
    'attachments': list([
    ]),
    'business_id': '',
    'content': dict({
    }),
    'content_schema_url': None,
    'created_at': '2020-06-01T03:00:00+03:00',
    'deletable': True,
    'delete_after': None,
    'document_language': None,
    'draft': False,
    'human_readable_type': dict({
    }),
    'id': '485af718-d9d1-46b9-ad7b-33ea054126e3',
    'locked_after': None,
    'metadata': dict({
    }),
    'service': 'service 218',
    'status': dict({
      'status_display_values': dict({
      }),
      'timestamp': '2020-06-01T03:00:00+03:00',
      'value': 'testing',
    }),
    'status_histories': list([
    ]),
    'tos_function_id': '81eee139736f4e52b046a1b27211c202',
    'tos_record_id': '0f499febb6414e1dafc93febca5ef312',
    'transaction_id': 'bd3fd958-cfeb-4ab1-bea4-5c058e8fee5c',
    'type': '',
    'updated_at': '2020-06-01T03:00:00+03:00',
  })
# ---
# name: test_retrieve_document_owner
  dict({
    'attachments': list([
    ]),
    'business_id': '',
    'content': dict({
    }),
    'content_schema_url': None,
    'created_at': '2020-06-01T03:00:00+03:00',
    'deletable': True,
    'delete_after': None,
    'document_language': None,
    'draft': False,
    'human_readable_type': dict({
    }),
    'id': '485af718-d9d1-46b9-ad7b-33ea054126e3',
    'locked_after': None,
    'metadata': dict({
    }),
    'service': 'service 217',
    'status': dict({
      'status_display_values': dict({
      }),
      'timestamp': '2020-06-01T03:00:00+03:00',
      'value': 'testing',
    }),
    'status_histories': list([
    ]),
    'tos_function_id': '81eee139736f4e52b046a1b27211c202',
    'tos_record_id': '0f499febb6414e1dafc93febca5ef312',
    'transaction_id': 'bd3fd958-cfeb-4ab1-bea4-5c058e8fee5c',
    'type': '',
    'updated_at': '2020-06-01T03:00:00+03:00',
  })
# ---
